source/slang/slang-ir-transform-params-to-constref.cpp

```diff
@@ -0,0 +1,325 @@
+        }
+
+        return true;
+    }
+
+    void rewriteParamUseSitesToSupportConstRefUsage(HashSet<IRParam*>& updatedParams)
+    {
+        // Traverse the uses of our updated params to rewrite them.
+        // Assume a `in` parameter has been converted to a `constref` parameter.
+        for (auto param : updatedParams)
+        {
+            traverseUses(
+                param,
+                [&](IRUse* use)
+                {
+                    auto user = use->getUser();
+                    switch (user->getOp())
+                    {
+                    case kIROp_FieldExtract:
+                        {
+                            // Transform the IRFieldExtract into a IRFieldAddress
+                            auto fieldExtract = as<IRFieldExtract>(use->getUser());
+                            builder.setInsertBefore(fieldExtract);
+                            auto fieldAddr = builder.emitFieldAddress(
+                                fieldExtract->getBase(),
+                                fieldExtract->getField());
+                            auto loadInst = builder.emitLoad(fieldAddr);
+                            fieldExtract->replaceUsesWith(loadInst);
+                            fieldExtract->removeAndDeallocate();
+                            break;
+                        }
+                    case kIROp_GetElement:
+                        {
+                            // Transform the IRGetElement into a IRGetElementPtr
+                            auto getElement = as<IRGetElement>(use->getUser());
+
+                            builder.setInsertBefore(getElement);
+                            auto elemAddr = builder.emitElementAddress(
+                                getElement->getBase(),
+                                getElement->getIndex());
+                            auto loadInst = builder.emitLoad(elemAddr);
+                            getElement->replaceUsesWith(loadInst);
+                            getElement->removeAndDeallocate();
+                            break;
+                        }
+                    default:
+                        {
+                            // Insert a load before the user and replace the user with the load
+                            builder.setInsertBefore(user);
+                            auto loadInst = builder.emitLoad(param);
+                            use->set(loadInst);
+                            break;
+                        }
+                    }
+                });
+        }
+    }
+
+    // Check if an instruction represents an addressable value
+    IRInst* getAddressable(IRInst* inst)
+    {
+        if (!inst)
+            return nullptr;
+
+        switch (inst->getOp())
+        {
+        case kIROp_Var:
+        case kIROp_GlobalVar:
+        case kIROp_FieldAddress:
+        case kIROp_GetElementPtr:
+            return inst;
+        case kIROp_GlobalParam:
+            return builder.emitLoad(inst);
```

Comment:
I am very confused what this function is doing. Why emit a load here for global param?